### PR TITLE
Fix normalization of cons-lists into bracketed lists

### LINKED ISSUE #1912: Cons-lists no longer reformatted to "traditional" list
**Is your feature request related to a problem? Please describe.**

At 2d6fc3f11, running `ocamlformat --profile=janestreet --enable-outside-detected-project` on the following input:

```ocaml
let x = 1 :: 2 :: 3 :: 4 :: []
```

makes no change.

**Describe the solution you'd like**

Previous versions of ocamlformat (at least, 6620833ca, but presumably earlier as well) would reformat this to

```ocaml
let x = [1; 2; 3; 4]
```

I think we'd definitely prefer this styling in general. This also occurs with lists that would be wrapped, e.g

```ocaml
let x = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" :: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb" :: "ccccccccccccccccccccccccccc" :: []
```

becoming

```ocaml
let x =
  "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" 
  :: "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
  :: "ccccccccccccccccccccccccccc"
  :: []
```

rather than

```ocaml
let x =
  [ "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" 
  ; "bbbbbbbbbbbbbbbbbbbbbbbbbbbbbbbb"
  ; "ccccccccccccccccccccccccccc"
  ]
```

### PR DESCRIPTION
**Pull request by gpetiot**
Fix #1912, here is the diff of test_branch.sh for the janestreet profile:

<details>

```diff
diff --git a/ocamldoc/odoc.ml b/ocamldoc/odoc.ml
index 724752682..adc363195 100644
--- a/ocamldoc/odoc.ml
+++ b/ocamldoc/odoc.ml
@@ -23,7 +23,7 @@ let arg_list = Array.to_list Sys.argv
 
 let plugins, paths =
   let rec iter (files, incs) = function
-    | [] | _ :: [] -> List.rev files, List.rev incs
+    | [] | [ _ ] -> List.rev files, List.rev incs
     | "-g" :: file :: q
       when Filename.check_suffix file "cmo"
            || Filename.check_suffix file "cma"
diff --git a/ocamldoc/odoc_ast.ml b/ocamldoc/odoc_ast.ml
index 5baecd5c8..336dcefb7 100644
--- a/ocamldoc/odoc_ast.ml
+++ b/ocamldoc/odoc_ast.ml
@@ -287,7 +287,7 @@ functor
           Odoc_parameter.Tuple ([], Odoc_env.subst_type env pattern_param.pat_type)
         in
         [ parameter ]
-      | { c_lhs = pattern_param; c_rhs = func_body } :: [] ->
+      | [ { c_lhs = pattern_param; c_rhs = func_body } ] ->
         let parameter =
           tt_param_info_from_pattern
             env
@@ -457,7 +457,7 @@ functor
                 }
             in
             [ new_param ]
-          | { c_lhs = pattern_param; c_rhs = body } :: [] ->
+          | [ { c_lhs = pattern_param; c_rhs = body } ] ->
             (* if this is the first call to the function, this is the first parameter and we skip it *)
             if not first
             then (
diff --git a/ocamldoc/odoc_html.ml b/ocamldoc/odoc_html.ml
index 4bbdfa37f..71f2894b5 100644
--- a/ocamldoc/odoc_html.ml
+++ b/ocamldoc/odoc_html.ml
@@ -698,7 +698,7 @@ class virtual info =
     method html_of_raised_exceptions b l =
       match l with
       | [] -> ()
-      | (s, t) :: [] ->
+      | [ (s, t) ] ->
         bp b "<li><b>%s</b> <code>%s</code> " Odoc_messages.raises s;
         self#html_of_text b t;
         bs b "</li>\n"
@@ -727,7 +727,7 @@ class virtual info =
     method html_of_sees b l =
       match l with
       | [] -> ()
-      | see :: [] ->
+      | [ see ] ->
         bp b "<li><b>%s</b> " Odoc_messages.see_also;
         self#html_of_see b see;
         bs b "</li>\n"
@@ -1905,7 +1905,7 @@ module Generator = struct
       method html_of_parameter_description b p =
         match Parameter.names p with
         | [] -> ()
-        | name :: [] ->
+        | [ name ] ->
           (* Only one name, no need for label for the description. *)
           (match Parameter.desc_by_name p name with
           | None -> ()
@@ -2398,7 +2398,7 @@ module Generator = struct
         fun f_generate l ->
           let rec iter pre_opt = function
             | [] -> ()
-            | ele :: [] -> f_generate pre_opt None ele
+            | [ ele ] -> f_generate pre_opt None ele
             | ele1 :: ele2 :: q ->
               f_generate pre_opt (Some ele2) ele1;
index 4bbdfa37f..71f2894b5 100644
--- a/ocamldoc/odoc_html.ml
+++ b/ocamldoc/odoc_html.ml
@@ -698,7 +698,7 @@ class virtual info =
     method html_of_raised_exceptions b l =
       match l with
       | [] -> ()
-      | (s, t) :: [] ->
+      | [ (s, t) ] ->
         bp b "<li><b>%s</b> <code>%s</code> " Odoc_messages.raises s;
         self#html_of_text b t;
         bs b "</li>\n"
@@ -727,7 +727,7 @@ class virtual info =
     method html_of_sees b l =
       match l with
       | [] -> ()
-      | see :: [] ->
+      | [ see ] ->
         bp b "<li><b>%s</b> " Odoc_messages.see_also;
         self#html_of_see b see;
         bs b "</li>\n"
@@ -1905,7 +1905,7 @@ module Generator = struct
       method html_of_parameter_description b p =
         match Parameter.names p with
         | [] -> ()
-        | name :: [] ->
+        | [ name ] ->
           (* Only one name, no need for label for the description. *)
           (match Parameter.desc_by_name p name with
           | None -> ()
@@ -2398,7 +2398,7 @@ module Generator = struct
         fun f_generate l ->
           let rec iter pre_opt = function
             | [] -> ()
-            | ele :: [] -> f_generate pre_opt None ele
+            | [ ele ] -> f_generate pre_opt None ele
             | ele1 :: ele2 :: q ->
               f_generate pre_opt (Some ele2) ele1;
               iter (Some ele1) (ele2 :: q)
diff --git a/ocamldoc/odoc_man.ml b/ocamldoc/odoc_man.ml
index ffc6a08f8..bb841137b 100644
--- a/ocamldoc/odoc_man.ml
+++ b/ocamldoc/odoc_man.ml
@@ -680,7 +680,7 @@ module Generator = struct
       method man_of_parameter_description b p =
         match Parameter.names p with
         | [] -> ()
-        | name :: [] ->
+        | [ name ] ->
           (* Only one name, no need for label for the description. *)
           (match Parameter.desc_by_name p name with
           | None -> ()
diff --git a/ocamldoc/odoc_merge.ml b/ocamldoc/odoc_merge.ml
index f0d08f1e1..a34c7e2a2 100644
--- a/ocamldoc/odoc_merge.ml
+++ b/ocamldoc/odoc_merge.ml
@@ -910,7 +910,7 @@ let merge merge_options modules_list =
       | [] ->
         (* no other module to merge with *)
         m :: iter l_others
-      | m2 :: [] ->
+      | [ m2 ] ->
         (* we can merge m with m2 if there is an implementation
                 and an interface.*)
         let f b = if !Odoc_global.inverse_merge_ml_mli then not b else b in
diff --git a/ocamldoc/odoc_misc.ml b/ocamldoc/odoc_misc.ml
index b8b34bd66..0aa4a47b3 100644
--- a/ocamldoc/odoc_misc.ml
+++ b/ocamldoc/odoc_misc.ml
@@ -156,7 +156,7 @@ let string_of_since_opt s_opt =
 let string_of_raised_exceptions l =
   match l with
   | [] -> ""
-  | (s, t) :: [] -> Odoc_messages.raises ^ " " ^ s ^ " " ^ string_of_text t ^ "\n"
+  | [ (s, t) ] -> Odoc_messages.raises ^ " " ^ s ^ " " ^ string_of_text t ^ "\n"
   | _ ->
     Odoc_messages.raises
     ^ "\n"
@@ -180,7 +180,7 @@ let string_of_see (see_ref, t) =
 let string_of_sees l =
   match l with
   | [] -> ""
-  | see :: [] -> Odoc_messages.see_also ^ " " ^ string_of_see see ^ " \n"
+  | [ see ] -> Odoc_messages.see_also ^ " " ^ string_of_see see ^ " \n"
   | _ ->
     Odoc_messages.see_also
     ^ "\n"
diff --git a/ocamldoc/odoc_sig.ml b/ocamldoc/odoc_sig.ml
index a6451aaaf..dd4088ee5 100644
--- a/ocamldoc/odoc_sig.ml
+++ b/ocamldoc/odoc_sig.ml
@@ -226,7 +226,7 @@ functor
       let doc p pos_end ld =
         let rec f = function
           | [] -> []
-          | ld :: [] ->
+          | [ ld ] ->
             let name = p.name ld in
             let pos = p.end_ ld in
             let _, comment_opt = just_after_special pos pos_end in
@@ -311,7 +311,7 @@ functor
               | Otag ({ txt = "" }, _) :: _ ->
                 (* Fields with no name have been eliminated previously. *)
                 assert false
-              | Otag ({ txt = name }, ct) :: [] ->
+              | [ Otag ({ txt = name }, ct) ] ->
                 let pos = Loc.ptyp_end ct in
                 let _, comment_opt = just_after_special pos pos_end in
                 [ name, comment_opt ]
@@ -335,7 +335,7 @@ functor
           let open Parsetree in
           match cons_core_type_list_list with
           | [] -> 0, acc
-          | pcd :: [] ->
+          | [ pcd ] ->
             let acc = Record.(inline_doc parsetree) pcd @ acc in
             let len, comment_opt = just_after_special (Loc.pcd_end pcd) pos_limit in
             len, List.rev @@ ((pcd.pcd_name.txt, comment_opt) :: acc)
diff --git a/ocamldoc/odoc_texi.ml b/ocamldoc/odoc_texi.ml
index bb3d8cc00..4e2eef43f 100644
--- a/ocamldoc/odoc_texi.ml
+++ b/ocamldoc/odoc_texi.ml
@@ -470,7 +470,7 @@ module Generator = struct
       method! text_of_raised_exceptions =
         function
         | [] -> []
-        | (s, t) :: [] ->
+        | [ (s, t) ] ->
           [ linebreak; Bold [ Raw Odoc_messages.raises ]; Raw " "; Code s; Raw " " ]
           @ t
           @ [ Newline ]
diff --git a/ocamldoc/odoc_to_text.ml b/ocamldoc/odoc_to_text.ml
index 42276c4f6..bb380c3d4 100644
--- a/ocamldoc/odoc_to_text.ml
+++ b/ocamldoc/odoc_to_text.ml
@@ -69,7 +69,7 @@ class virtual info =
     method text_of_raised_exceptions l =
       match l with
       | [] -> []
-      | (s, t) :: [] ->
+      | [ (s, t) ] ->
         [ Bold [ Raw Odoc_messages.raises ]; Raw " "; Code s; Raw " " ] @ t @ [ Newline ]
       | _ ->
         [ Bold [ Raw Odoc_messages.raises ]
@@ -91,7 +91,7 @@ class virtual info =
     method text_of_sees l =
       match l with
       | [] -> []
-      | see :: [] ->
+      | [ see ] ->
         (Bold [ Raw Odoc_messages.see_also ] :: Raw " " :: self#text_of_see see)
         @ [ Newline ]
       | _ ->
@@ -365,7 +365,7 @@ class virtual to_text =
     method text_of_parameter_description p =
       match Parameter.names p with
       | [] -> []
-      | name :: [] ->
+      | [ name ] ->
         (* Only one name, no need for label for the description. *)
         (match Parameter.desc_by_name p name with
         | None -> []
diff --git a/stdlib/bytes.ml b/stdlib/bytes.ml
index 36b5bb8ce..18941b760 100644
--- a/stdlib/bytes.ml
+++ b/stdlib/bytes.ml
@@ -139,13 +139,13 @@ let ensure_ge (x : int) y = if x >= y then x else invalid_arg "Bytes.concat"
 
 let rec sum_lengths acc seplen = function
   | [] -> acc
-  | hd :: [] -> length hd + acc
+  | [ hd ] -> length hd + acc
   | hd :: tl -> sum_lengths (ensure_ge (length hd + seplen + acc) acc) seplen tl
 ;;
 
 let rec unsafe_blits dst pos sep seplen = function
   | [] -> dst
-  | hd :: [] ->
+  | [ hd ] ->
     unsafe_blit hd 0 dst pos (length hd);
     dst
   | hd :: tl ->
       | [] -> []
-      | see :: [] ->
+      | [ see ] ->
         (Bold [ Raw Odoc_messages.see_also ] :: Raw " " :: self#text_of_see see
)
         @ [ Newline ]
       | _ ->
@@ -365,7 +365,7 @@ class virtual to_text =
     method text_of_parameter_description p =
       match Parameter.names p with
       | [] -> []
-      | name :: [] ->
+      | [ name ] ->
         (* Only one name, no need for label for the description. *)
         (match Parameter.desc_by_name p name with
         | None -> []
diff --git a/stdlib/bytes.ml b/stdlib/bytes.ml
index 36b5bb8ce..18941b760 100644
--- a/stdlib/bytes.ml
+++ b/stdlib/bytes.ml
@@ -139,13 +139,13 @@ let ensure_ge (x : int) y = if x >= y then x else invalid_
arg "Bytes.concat"
 
 let rec sum_lengths acc seplen = function
   | [] -> acc
-  | hd :: [] -> length hd + acc
+  | [ hd ] -> length hd + acc
   | hd :: tl -> sum_lengths (ensure_ge (length hd + seplen + acc) acc) seplen t
l
 ;;
 
 let rec unsafe_blits dst pos sep seplen = function
   | [] -> dst
-  | hd :: [] ->
+  | [ hd ] ->
     unsafe_blit hd 0 dst pos (length hd);
     dst
   | hd :: tl ->
           | [] -> 0, acc
-          | pcd :: [] ->
+          | [ pcd ] ->
             let acc = Record.(inline_doc parsetree) pcd @ acc in
             let len, comment_opt = just_after_special (Loc.pcd_end pcd) pos_lim
it in
             len, List.rev @@ ((pcd.pcd_name.txt, comment_opt) :: acc)
diff --git a/ocamldoc/odoc_texi.ml b/ocamldoc/odoc_texi.ml
index bb3d8cc00..4e2eef43f 100644
--- a/ocamldoc/odoc_texi.ml
+++ b/ocamldoc/odoc_texi.ml
@@ -470,7 +470,7 @@ module Generator = struct
       method! text_of_raised_exceptions =
         function
         | [] -> []
-        | (s, t) :: [] ->
+        | [ (s, t) ] ->
           [ linebreak; Bold [ Raw Odoc_messages.raises ]; Raw " "; Code s; Raw 
" " ]
           @ t
           @ [ Newline ]
diff --git a/ocamldoc/odoc_to_text.ml b/ocamldoc/odoc_to_text.ml
index 42276c4f6..bb380c3d4 100644
--- a/ocamldoc/odoc_to_text.ml
+++ b/ocamldoc/odoc_to_text.ml
@@ -69,7 +69,7 @@ class virtual info =
     method text_of_raised_exceptions l =
       match l with
       | [] -> []
-      | (s, t) :: [] ->
+      | [ (s, t) ] ->
         [ Bold [ Raw Odoc_messages.raises ]; Raw " "; Code s; Raw " " ] @ t @ [
 Newline ]
       | _ ->
         [ Bold [ Raw Odoc_messages.raises ]
@@ -91,7 +91,7 @@ class virtual info =
     method text_of_sees l =
       match l with
       | [] -> []
-      | see :: [] ->
+      | [ see ] ->
         (Bold [ Raw Odoc_messages.see_also ] :: Raw " " :: self#text_of_see see)
         @ [ Newline ]
       | _ ->
@@ -365,7 +365,7 @@ class virtual to_text =
     method text_of_parameter_description p =
       match Parameter.names p with
       | [] -> []
-      | name :: [] ->
+      | [ name ] ->
         (* Only one name, no need for label for the description. *)
         (match Parameter.desc_by_name p name with
         | None -> []
diff --git a/stdlib/bytes.ml b/stdlib/bytes.ml
index 36b5bb8ce..18941b760 100644
--- a/stdlib/bytes.ml
+++ b/stdlib/bytes.ml
@@ -139,13 +139,13 @@ let ensure_ge (x : int) y = if x >= y then x else invalid_arg "Bytes.concat"
 
 let rec sum_lengths acc seplen = function
   | [] -> acc
-  | hd :: [] -> length hd + acc
+  | [ hd ] -> length hd + acc
   | hd :: tl -> sum_lengths (ensure_ge (length hd + seplen + acc) acc) seplen tl
 ;;
 
 let rec unsafe_blits dst pos sep seplen = function
   | [] -> dst
-  | hd :: [] ->
+  | [ hd ] ->
     unsafe_blit hd 0 dst pos (length hd);
     dst
   | hd :: tl ->
diff --git a/stdlib/string.ml b/stdlib/string.ml
index 3c7c7adf3..845301ca1 100644
--- a/stdlib/string.ml
+++ b/stdlib/string.ml
@@ -51,13 +51,13 @@ let ensure_ge (x : int) y = if x >= y then x else invalid_arg "String.concat"
 
 let rec sum_lengths acc seplen = function
   | [] -> acc
-  | hd :: [] -> length hd + acc
+  | [ hd ] -> length hd + acc
   | hd :: tl -> sum_lengths (ensure_ge (length hd + seplen + acc) acc) seplen tl
 ;;
 
 let rec unsafe_blits dst pos sep seplen = function
   | [] -> dst
-  | hd :: [] ->
+  | [ hd ] ->
     unsafe_blit hd 0 dst pos (length hd);
     dst
   | hd :: tl ->
diff --git a/testsuite/tests/typing-warnings/records.ml b/testsuite/tests/typing-warnings/records.ml
index d20372188..ac0c5a31a 100644
--- a/testsuite/tests/typing-warnings/records.ml
+++ b/testsuite/tests/typing-warnings/records.ml
@@ -933,8 +933,8 @@ module M = struct
 end
 
 let f (x : M.t) = { x with y = 'a' }
-let g (x : M.t) = { x with y = 'a' } :: []
-let h (x : M.t) = { x with y = 'a' } :: { x with y = 'b' } :: []
+let g (x : M.t) = [ { x with y = 'a' } ]
+let h (x : M.t) = [ { x with y = 'a' }; { x with y = 'b' } ]
 
 [%%expect
 {|
```

</details>

In the test I added the comments are a bit sketchy, but since the ( :: ) lists are rewritten as bracketed lists, the comments can be moved after the rewriting and their new position will be preserved.

I'm not sure whether we should name this step "normalize", as the compiler's normalization transforms the bracketed lists into cons lists, and we used to call the opposite translation "sugarize", not sure it's important.

cc @cwong-ocaml @ceastlund for feedback on the janestreet codebase.

### CODE REVIEW

File: lib/Extended_ast.mli [hunk 1]
```diff
@@ -24,7 +24,7 @@ type 'a t =
   | Expression : expression t
 
 module Parse : sig
-  val ast : 'a t -> Lexing.lexbuf -> 'a
+  val ast : 'a t -> conf:Conf.t -> Lexing.lexbuf -> 'a
```
Comment (Julow) [inline lib/Extended_ast.mli:27]: Passing the `Conf.t` record entirely for one option feels a bit out of scope.
I think it would be clearer with a smaller "conf" record defined in this module.

**Review by Julow — COMMENTED** [review]
I feel like this will be requested by users of other profiles as this was a feature of ocamlformat.

Perhaps we can consider #1694 (which was initially fixing comments) shouldn't have removed that feature and enable again this normalization for every profiles.

**Comment by gpetiot** [comment]
> I feel like this will be requested by users of other profiles as this was a feature of ocamlformat.
> 
> Perhaps we can consider #1694 (which was initially fixing comments) shouldn't have removed that feature and enable again this normalization for every profiles.

It's fine by me, thoughts @jberdine for the `ocamlformat` profile, and @samoht for the default profile?

**Comment by gpetiot** [comment]
Since this change (#1694) has been introduced between 0.19.0 and now (0.20.0 not released yet) I think it's better to apply the change of this PR across all profiles (default and ocamlformat included then) to minimize the diff with 0.19.0.
